src/storage/src/hummock/sstable/delete_range_aggregator.rs

```diff
@@ -203,24 +225,45 @@ impl CompactionDeleteRanges {
             return vec![];
         }
 
-        let mut tombstones = Vec::with_capacity(self.delete_tombstones.len());
-        for tombstone in &self.delete_tombstones {
-            let mut candidate = tombstone.clone();
-            if !smallest_user_key.is_empty()
-                && smallest_user_key.gt(&candidate.start_user_key.as_ref())
-            {
-                candidate.start_user_key = smallest_user_key.to_vec();
-            }
-            if !largest_user_key.is_empty() && largest_user_key.lt(&candidate.end_user_key.as_ref())
-            {
-                candidate.end_user_key = largest_user_key.to_vec();
+        let mut monotonic_events = Vec::with_capacity(self.events.len());
+        let mut epochs = BTreeSet::new();
+        let mut idx = 0;
+        while idx < self.events.len() {
+            if self.events[idx].0.as_ref().gt(&smallest_user_key) {
+                if let Some(epoch) = epochs.first() {
+                    monotonic_events.push(MonotonicDeleteEvent {
+                        event_key: smallest_user_key.to_vec(),
+                        new_epoch: *epoch,
+                        is_exclusive: false,
+                    });
+                }
+                break;
             }
-            if candidate.start_user_key < candidate.end_user_key {
-                tombstones.push(candidate);
+            apply_event(&mut epochs, &self.events[idx]);
+            idx += 1;
+        }
+        while idx < self.events.len() {
+            // TODO: replace it with Bound
+            if !largest_user_key.is_empty() && self.events[idx].0.as_ref().ge(&largest_user_key) {
```

Comment:
If the first event is greater than `largest_user_key`, we do not want to add an extra `MonotonicDeleteEvent` because it will create an extra delete range iter in `iter()`.